compiler/rustc_resolve/src/late/lifetimes.rs

```diff
@@ -963,45 +1129,111 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
         let should_pop_missing_lt = self.is_trait_ref_fn_scope(trait_ref);
 
         let trait_ref_hack = take(&mut self.trait_ref_hack);
-        if !trait_ref_hack
-            || trait_ref
+        let next_early_index = self.next_early_index();
+        let mut named_late_bound_vars = 0;
+        let lifetimes: FxHashMap<hir::ParamName, Region> = trait_ref
+            .bound_generic_params
+            .iter()
+            .filter_map(|param| match param.kind {
+                GenericParamKind::Lifetime { .. } => {
+                    let late_bound_idx = named_late_bound_vars;
+                    named_late_bound_vars += 1;
+                    Some(Region::late(late_bound_idx, &self.tcx.hir(), param))
+                }
+                _ => None,
+            })
+            .collect();
+        // See note on `trait_ref_hack`. If `for<..>` has been defined in both
+        // the outer and inner part of the trait ref, emit an error.
+        let has_lifetimes = trait_ref.bound_generic_params.iter().any(|param| match param.kind {
+            GenericParamKind::Lifetime { .. } => true,
+            _ => false,
+        });
+        if trait_ref_hack && has_lifetimes {
+            struct_span_err!(
+                self.tcx.sess,
+                trait_ref.span,
+                E0316,
+                "nested quantification of lifetimes"
+            )
+            .emit();
+        }
+        let (binders, named_late_bound_vars) = if trait_ref_hack {
+            // Since `trait_ref_hack` is true, we *know* there is a `Scope::Binder` above.
+            let (mut binders, mut named_late_bound_vars) = {
+                let mut scope = self.scope;
+                loop {
+                    match *scope {
+                        Scope::Body { s, .. }
+                        | Scope::Elision { s, .. }
+                        | Scope::ObjectLifetimeDefault { s, .. } => {
+                            scope = s;
+                        }
+                        Scope::Root | Scope::TraitRefHackInner { .. } => bug!(),
+                        Scope::Binder { ref binders, named_late_bound_vars, .. } => {
+                            break (binders.clone(), named_late_bound_vars);
+                        }
+                    }
+                }
+            };
+
+            binders.extend(trait_ref.bound_generic_params.iter().filter_map(
```

Comment:
Unfortunately, even though this is a semantic error, we have to handle it gracefully